dbms/src/DataStreams/HashJoinProbeBlockInputStream.cpp

```diff
@@ -78,23 +94,65 @@ Block HashJoinProbeBlockInputStream::readImpl()
         Block result_block = getOutputBlock();
         squashing_transform.appendBlock(result_block);
     }
-    return squashing_transform.getFinalOutputBlock();
+    auto ret = squashing_transform.getFinalOutputBlock();
+    return ret;
+}
+
+void HashJoinProbeBlockInputStream::readSuffixImpl()
+{
+    LOG_DEBUG(log, "Finish join probe, total output rows {}, joined rows {}, non joined rows {}", joined_rows + non_joined_rows, joined_rows, non_joined_rows);
 }
 
 Block HashJoinProbeBlockInputStream::getOutputBlock()
 {
-    if (probe_process_info.all_rows_joined_finish)
+    while (true)
     {
-        Block block = children.back()->read();
-        if (!block)
+        switch (status)
+        {
+        case ProbeStatus::PROBE:
         {
+            if (probe_process_info.all_rows_joined_finish)
+            {
+                Block block = children.back()->read();
+                if (!block)
+                {
+                    join->finishOneProbe();
+                    if (join->needReturnNonJoinedData())
+                        status = ProbeStatus::WAIT_FOR_READ_NON_JOINED_DATA;
+                    else
+                        status = ProbeStatus::FINISHED;
+                    break;
```

Comment:
Is the `break` here to jump out of `switch(status)` or out of `while(true)`?

Reply:
I think it will break the `swith(status)`

Reply:
ok